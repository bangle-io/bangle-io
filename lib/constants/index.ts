// This package should not import any other package and should be mostly
// primitives.
export * from './core/core-dialogs';
export * from './core/core-operations';
export * from './core/core-palettes';
export * from './editor';
export * from './help-fs';
export * from './search';
export * from './store';
export * from './ui';
export * from './worker';
export * from './workspace';
export * from './write-disk';
