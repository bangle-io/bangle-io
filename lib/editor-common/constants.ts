import { PluginKey } from '@bangle.dev/pm';

import type { IntersectionObserverPluginState } from '@bangle.io/pm-plugins';
import type { EditorPluginMetadata } from '@bangle.io/shared-types';

export const intersectionObserverPluginKey =
  new PluginKey<IntersectionObserverPluginState>(
    'core-editor_intersectionObserverPlugin',
  );

export const EditorPluginMetadataKey = new PluginKey<EditorPluginMetadata>(
  'EditorPluginMetadataKey',
);
export const menuKey = new PluginKey('menuKey');

export const searchPluginKey = new PluginKey('searchPluginKey');
