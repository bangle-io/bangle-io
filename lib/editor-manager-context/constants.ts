import { SliceKey } from '@bangle.io/create-store';

import type { EditorManagerAction, EditorSliceState } from './types';

export const editorManagerSliceKey = new SliceKey<
  EditorSliceState,
  EditorManagerAction
>('editor-manager-slice');

// The time to wait before auto focusing any newly
// mounted editor
export const FOCUS_EDITOR_ON_LOAD_COOLDOWN = 400;
