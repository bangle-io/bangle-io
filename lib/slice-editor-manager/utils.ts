import type { BangleEditor } from '@bangle.dev/core';
import type { EditorView } from '@bangle.dev/pm';

import { MAX_OPEN_EDITORS } from '@bangle.io/constants';
import { getEditorPluginMetadata } from '@bangle.io/editor-common';
import { z } from '@bangle.io/nsm-3';
import type { EditorIdType, WsPath } from '@bangle.io/shared-types';
import { findWrappingScrollable } from '@bangle.io/utils';

import type { EditorSliceState } from './types';

// In PM selection.toJSON() is of different types based on the selection type
// we only care about serializing and this type is good enough.
export type SelectionJson = z.infer<typeof selectionJsonSchema>;
// WARNING: before changing this schema, make sure to update the
//  make sure to update the manual check in `calculateSelection`
export const selectionJsonSchema = z.record(
  z.union([z.undefined(), z.string(), z.number(), z.null()]),
);

export const calculateSelection = (
  editorId: EditorIdType,
  editor: BangleEditor,
): {
  wsPath: WsPath;
  editorId: EditorIdType;
  selectionJson: SelectionJson;
} => {
  const selection = editor.view.state.selection;

  const json = selection.toJSON();

  for (const value of Object.values(json)) {
    if (
      typeof value !== 'number' &&
      typeof value !== 'string' &&
      value != null
    ) {
      throw new Error(`Invalid editor selection json value type: ${json}`);
    }
  }

  return {
    wsPath: getEditorPluginMetadata(editor.view.state).wsPath,
    editorId: editorId,
    selectionJson: json as SelectionJson,
  };
};

export const calculateScrollPosition = (
  editorId: EditorIdType,
  view: EditorView,
) => {
  const top = getScrollParentElement(editorId)?.scrollTop;

  if (typeof top === 'number') {
    return {
      wsPath: getEditorPluginMetadata(view.state).wsPath,
      editorId: editorId,
      scrollPosition: top,
    };
  }

  return undefined;
};

export function* getEachEditorIterable(
  editorManagerState: EditorSliceState,
): Iterable<{
  editor: BangleEditor | undefined;
  editorId: NonNullable<EditorIdType>;
}> {
  for (const [index, editor] of editorManagerState.mainEditors.entries()) {
    yield { editor, editorId: index };
  }
}

export function isValidEditorId(editorId: EditorIdType): boolean {
  return editorId >= 0 && editorId < MAX_OPEN_EDITORS;
}

export function assertValidEditorId(
  editorId: EditorIdType,
): asserts editorId is EditorIdType {
  if (!isValidEditorId(editorId)) {
    throw new Error('editorId is out of range or invalid');
  }
}

export function getScrollParentElement(editorId: EditorIdType) {
  const editor = document.querySelector(
    '.B-editor-container_editor-' + editorId,
  );

  if (editor) {
    return findWrappingScrollable(editor);
  }

  return undefined;
}
