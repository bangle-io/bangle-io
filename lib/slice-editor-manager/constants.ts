import { MAX_OPEN_EDITORS } from '@bangle.io/constants';
import { createEmptyArray } from '@bangle.io/mini-js-utils';
import { checkWidescreen } from '@bangle.io/utils';

import { OpenedEditorsConfig } from './opened-editors-config';
import type { EditorSliceState } from './types';

// The time to wait before auto focusing any newly
// mounted editor
export const FOCUS_EDITOR_ON_LOAD_COOLDOWN = 1500;

export const initialEditorSliceState: EditorSliceState = {
  focusedEditorId: undefined,
  mainEditors: createEmptyArray(MAX_OPEN_EDITORS),
  editorConfig: OpenedEditorsConfig.fromJsonObj({
    selections: [],
    scrollPositions: [],
  }),
  primaryEditor: undefined,
  secondaryEditor: undefined,

  // We disable editing in mobile devices by default.
  // TODO: move this to a config stating the widescreen status
  editingAllowed: checkWidescreen(),

  editorOpenOrder: [],
  disableEditingCounter: undefined,
  searchQuery: undefined,
};
