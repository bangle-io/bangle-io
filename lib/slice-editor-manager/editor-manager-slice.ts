import { isMobile } from '@bangle.io/config';
import { MAX_OPEN_EDITORS } from '@bangle.io/constants';
import { Slice } from '@bangle.io/create-store';
import { assertActionName, createEmptyArray } from '@bangle.io/utils';

import { editorManagerSliceKey } from './constants';
import {
  focusEditorEffect,
  initialSelectionEffect,
  trimWhiteSpaceEffect,
  watchEditorScrollEffect,
} from './effects';
import { OpenedEditorsConfig } from './opened-editors-config';
import type { EditorManagerAction, EditorSliceState } from './types';
import { calculateScrollPosition, calculateSelection } from './utils';

const LOG = false;
let log = LOG ? console.log.bind(console, 'editorManagerSlice') : () => {};

export const JSON_SCHEMA_VERSION = 'editor-slice/2';

export const initialEditorSliceState: EditorSliceState = {
  focusedEditorId: undefined,
  editors: createEmptyArray(MAX_OPEN_EDITORS),
  editorConfig: OpenedEditorsConfig.fromJsonObj({
    selections: [],
    scrollPositions: [],
  }),
  primaryEditor: undefined,
  secondaryEditor: undefined,
  // for now disabling editing for mobile by default
  // as it causes a lot of jumps due to keyboard
  editingAllowed: !isMobile,
};

const applyState = (
  action: EditorManagerAction,
  state: EditorSliceState,
): EditorSliceState => {
  switch (action.name) {
    case 'action::@bangle.io/slice-editor-manager:toggle-editing': {
      return {
        ...state,
        editingAllowed: !state.editingAllowed,
      };
    }
    case 'action::@bangle.io/slice-editor-manager:set-editor': {
      const { editorId, editor } = action.value;
      if (editorId >= MAX_OPEN_EDITORS) {
        throw new Error('editorId is out of range');
      }

      const newEditors: EditorSliceState['editors'] = [...state.editors];

      newEditors[editorId] = editor;

      return {
        ...state,
        editors: newEditors,
      };
    }

    case 'action::@bangle.io/slice-editor-manager:on-focus-update': {
      const editorId = action.value.editorId;
      if (typeof editorId === 'number' && editorId >= MAX_OPEN_EDITORS) {
        throw new Error('editorId is out of range');
      }

      return {
        ...state,
        focusedEditorId: editorId,
      };
    }

    case 'action::@bangle.io/slice-editor-manager:update-scroll-position': {
      const { editorId, wsPath, scrollPosition } = action.value;

      const newEditorConfig = state.editorConfig.updateScrollPosition(
        scrollPosition,
        wsPath,
        editorId,
      );

      return {
        ...state,
        editorConfig: newEditorConfig,
      };
    }

    case 'action::@bangle.io/slice-editor-manager:update-initial-selection-json': {
      const { editorId, wsPath, selectionJson } = action.value;

      const newEditorConfig = state.editorConfig.updateSelection(
        selectionJson,
        wsPath,
        editorId,
      );

      return {
        ...state,
        editorConfig: newEditorConfig,
      };
    }

    default:
      return state;
  }
};

export function editorManagerSlice(): Slice<
  EditorSliceState,
  EditorManagerAction
> {
  assertActionName('@bangle.io/slice-editor-manager', editorManagerSliceKey);

  return new Slice({
    key: editorManagerSliceKey,
    state: {
      init() {
        return initialEditorSliceState;
      },
      apply(action, state) {
        const newState = applyState(action, state);
        if (newState === state) {
          return state;
        }

        // derived state
        newState.primaryEditor = newState.editors[0];
        newState.secondaryEditor = newState.editors[1];

        return newState;
      },
      stateFromJSON(_, value: any) {
        if (!value || value.version !== JSON_SCHEMA_VERSION) {
          return initialEditorSliceState;
        }

        const data = value.data;

        const sliceState: EditorSliceState = Object.assign(
          {},
          initialEditorSliceState,
          {
            editorConfig: OpenedEditorsConfig.fromJsonObj(data.editorConfig),
            focusedEditorId: data.focusedEditorId,
          },
        );
        return sliceState;
      },

      stateToJSON(sliceState) {
        let newEditorConfig = sliceState.editorConfig;
        for (let i = 0; i < MAX_OPEN_EDITORS; i++) {
          const editor = sliceState.editors[i];
          if (editor) {
            const { editorId, selectionJson, wsPath } = calculateSelection(
              i,
              editor,
            );

            const scroll = calculateScrollPosition(i, editor)?.scrollPosition;
            newEditorConfig = newEditorConfig.updateSelection(
              selectionJson,
              wsPath,
              editorId,
            );
            newEditorConfig = newEditorConfig.updateScrollPosition(
              scroll,
              wsPath,
              editorId,
            );
          }
        }

        return {
          version: JSON_SCHEMA_VERSION,
          data: {
            ...initialEditorSliceState,
            editors: [],
            focusedEditorId: sliceState.focusedEditorId,
            editorConfig: newEditorConfig.toJsonObj(),
            primaryEditor: undefined,
            secondaryEditor: undefined,
          },
        };
      },
    },
    sideEffect: [
      initialSelectionEffect,
      focusEditorEffect,
      watchEditorScrollEffect,
      trimWhiteSpaceEffect,
    ],
  });
}
