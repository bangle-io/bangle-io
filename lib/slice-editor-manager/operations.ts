import type { BangleEditor } from '@bangle.dev/core';
import type {
  EditorState,
  EditorView,
  Node,
  Transaction,
} from '@bangle.dev/pm';
import { Selection } from '@bangle.dev/pm';

import {
  PRIMARY_EDITOR_INDEX,
  SECONDARY_EDITOR_INDEX,
} from '@bangle.io/constants';
import type { AppState } from '@bangle.io/create-store';
import { getScrollParentElement } from '@bangle.io/utils';

import { editorManagerSliceKey } from './constants';
import type { EditorDispatchType, EditorIdType } from './types';
import {
  calculateScrollPosition,
  calculateSelection,
  getEachEditorIterable,
} from './utils';

export function toggleEditing() {
  return editorManagerSliceKey.op((_, dispatch) => {
    dispatch({
      name: 'action::@bangle.io/slice-editor-manager:toggle-editing',
    });
  });
}

export function isEditingAllowed() {
  return editorManagerSliceKey.queryOp((state) => {
    return editorManagerSliceKey.getSliceStateAsserted(state).editingAllowed;
  });
}

export function getPrimaryEditor() {
  return editorManagerSliceKey.queryOp((state) => {
    return getEditor(PRIMARY_EDITOR_INDEX)(state);
  });
}

export function focusPrimaryEditor() {
  return editorManagerSliceKey.queryOp((state) => {
    return focusEditor(PRIMARY_EDITOR_INDEX)(state);
  });
}

export function getSecondaryEditor() {
  return editorManagerSliceKey.queryOp((state) => {
    return getEditor(SECONDARY_EDITOR_INDEX)(state);
  });
}

export function focusSecondaryEditor() {
  return editorManagerSliceKey.queryOp((state) => {
    return focusEditor(SECONDARY_EDITOR_INDEX)(state);
  });
}

export function getEditor(editorId: EditorIdType) {
  return (state: AppState): BangleEditor | undefined => {
    if (editorId == null) {
      return undefined;
    }

    return editorManagerSliceKey.getSliceState(state)?.mainEditors[editorId];
  };
}

export function focusEditor(editorId: EditorIdType = PRIMARY_EDITOR_INDEX) {
  return (state: AppState): boolean => {
    const editor = getEditor(editorId)(state);

    if (editor) {
      editor.focusView();

      return true;
    }

    return false;
  };
}
export function updateFocusedEditor(
  editorId: EditorIdType = PRIMARY_EDITOR_INDEX,
) {
  return (state: AppState, dispatch: EditorDispatchType): boolean => {
    dispatch({
      name: 'action::@bangle.io/slice-editor-manager:on-focus-update',
      value: {
        editorId,
      },
    });

    return true;
  };
}

export function updateInitialSelection(editorId: EditorIdType) {
  return (state: AppState, dispatch: EditorDispatchType): boolean => {
    const editor = getEditor(editorId)(state);

    if (!editor || typeof editorId !== 'number') {
      return false;
    }

    const value = calculateSelection(editorId, editor);

    if (value) {
      dispatch({
        name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
        value: value,
      });

      return true;
    }

    return false;
  };
}

export function forEachEditor(
  cb: (editor: BangleEditor | undefined, index: number) => void,
) {
  return (state: AppState): void => {
    const editorManagerState = editorManagerSliceKey.getSliceState(state);

    if (!editorManagerState) {
      return;
    }

    for (const { editor, editorId } of getEachEditorIterable(
      editorManagerState,
    )) {
      cb(editor, editorId);
    }
  };
}

export function getEditorView(editorId: EditorIdType) {
  return (state: AppState): EditorView | undefined => {
    if (editorId == null) {
      return undefined;
    }
    let editor = getEditor(editorId)(state);

    if (!editor) {
      return undefined;
    }

    return editor.view;
  };
}

export function getEditorState(editorId: EditorIdType) {
  return (state: AppState): EditorState | undefined => {
    return getEditorView(editorId)(state)?.state;
  };
}

export function geEditorScrollPosition(editorId: EditorIdType, wsPath: string) {
  return (state: AppState): number | undefined => {
    const sliceState = editorManagerSliceKey.getSliceState(state);

    if (!sliceState) {
      return undefined;
    }

    return sliceState.editorConfig.getScrollPosition(wsPath, editorId);
  };
}

export function updateScrollPosition(editorId: EditorIdType) {
  return (state: AppState, dispatch: EditorDispatchType): boolean => {
    const editor = getEditor(editorId)(state);

    if (!editor || typeof editorId !== 'number') {
      return false;
    }

    const result = calculateScrollPosition(editorId, editor);

    if (result) {
      dispatch({
        name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
        value: result,
      });

      return true;
    }

    return false;
  };
}

export function setEditorReady(
  editorId: EditorIdType,
  wsPath: string,
  editor: BangleEditor,
) {
  return (state: AppState, dispatch: EditorDispatchType): boolean => {
    if (typeof editorId !== 'number') {
      return false;
    }

    const scrollParent = getScrollParentElement(editorId);
    const pos = geEditorScrollPosition(editorId, wsPath)(state);

    if (typeof pos === 'number' && scrollParent) {
      scrollParent.scrollTop = pos;
    }

    dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor,
        editorId,
      },
    });

    return true;
  };
}

export function setEditorUnmounted(
  editorId: EditorIdType,
  editor: BangleEditor,
) {
  return (state: AppState, dispatch: EditorDispatchType): boolean => {
    if (typeof editorId !== 'number') {
      return false;
    }

    // make sure we are unsetting the correct editor
    if (getEditor(editorId)(state) === editor) {
      dispatch({
        name: 'action::@bangle.io/slice-editor-manager:set-editor',
        value: {
          editor: undefined,
          editorId,
        },
      });

      return true;
    }

    return false;
  };
}
// Gets the editor selection saved in the slice state
export function getInitialSelection(
  editorId: EditorIdType,
  wsPath: string,
  doc: Node,
) {
  return (state: AppState): Selection | undefined => {
    const sliceState = editorManagerSliceKey.getSliceState(state);

    if (!sliceState) {
      return undefined;
    }

    const initialSelection: any = sliceState.editorConfig.getSelection(
      wsPath,
      editorId,
    );

    if (initialSelection) {
      let selection =
        Math.max(initialSelection?.anchor, initialSelection?.head) >=
        doc.content.size
          ? Selection.atEnd(doc)
          : Selection.fromJSON(doc, initialSelection);
      let { from } = selection;

      if (from >= doc.content.size) {
        selection = Selection.atEnd(doc);
      } else {
        selection = Selection.near(doc.resolve(from));
      }

      return selection;
    }

    return undefined;
  };
}

// Returns true or false if a new editor was added
// or an editor was removed.
export function didSomeEditorChange(prevState: AppState) {
  return (state: AppState): boolean => {
    for (const { editor: currentEditor, editorId } of getEachEditorIterable(
      editorManagerSliceKey.getSliceStateAsserted(state),
    )) {
      const prevEditor = getEditor(editorId)(prevState);

      if (currentEditor === prevEditor) {
        continue;
      }

      return true;
    }

    return false;
  };
}

export function dispatchEditorCommand<T>(
  editorId: EditorIdType,
  cmdCallback: (
    state: EditorState,
    dispatch?: (tr: Transaction) => void,
    view?: EditorView,
  ) => T,
) {
  return (state: AppState): T | false => {
    const currentEditor = getEditor(editorId)(state);

    if (!currentEditor) {
      return false;
    }

    const view = currentEditor.view;

    return cmdCallback(view.state, view.dispatch, view);
  };
}
