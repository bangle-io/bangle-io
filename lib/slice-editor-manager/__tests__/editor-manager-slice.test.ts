import { Selection } from '@bangle.dev/pm';

import { ApplicationStore, AppState } from '@bangle.io/create-store';
import { createPMNode } from '@bangle.io/test-utils/create-pm-node';
import { getScrollParentElement } from '@bangle.io/utils';

import { geEditorScrollPosition, getInitialSelection } from '..';
import { editorManagerSliceKey } from '../constants';
import { editorManagerSlice } from '../editor-manager-slice';
import { getEditor } from '../operations';
import { createTestEditor } from './test-utils';

jest.mock('@bangle.io/utils', () => {
  const actual = jest.requireActual('@bangle.io/utils');
  return {
    ...actual,
    getScrollParentElement: jest.fn(() => {
      return undefined;
    }),
  };
});

let createStore = () =>
  ApplicationStore.create({
    scheduler: (cb) => {
      cb();
      return () => {};
    },
    storeName: 'editor-store',
    state: AppState.create({ slices: [editorManagerSlice()] }),
  });

describe('set editor action', () => {
  test('works', () => {
    let mockEditor = createTestEditor();
    let store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor,
        editorId: 0,
      },
    });

    expect(getEditor(0)(store.state)).toBe(mockEditor);

    expect(
      editorManagerSliceKey.getSliceState(store.state)?.primaryEditor,
    ).toBe(mockEditor);

    expect(
      editorManagerSliceKey.getSliceState(store.state)?.secondaryEditor,
    ).toBe(undefined);
  });

  test('setting out of range editorId', () => {
    let mockEditor = createTestEditor();
    let store = createStore();
    expect(() =>
      store.dispatch({
        name: 'action::@bangle.io/slice-editor-manager:set-editor',
        value: {
          editor: mockEditor,
          editorId: 10000,
        },
      }),
    ).toThrowError('editorId is out of range');
  });

  test('replacing editor', () => {
    let mockEditor = createTestEditor('test:first.md');
    let mockEditor2 = createTestEditor('test:second.md');

    let store = createStore();

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor,
        editorId: 0,
      },
    });

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor2,
        editorId: 0,
      },
    });

    expect(getEditor(0)(store.state)).toBe(mockEditor2);

    let mockEditor3 = createTestEditor('test:third.md');

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor3,
        editorId: 1,
      },
    });

    expect(
      editorManagerSliceKey.getSliceState(store.state)?.primaryEditor,
    ).toBe(mockEditor2);
    expect(
      editorManagerSliceKey.getSliceState(store.state)?.secondaryEditor,
    ).toBe(mockEditor3);
  });
});

describe('update focus action', () => {
  test('works', () => {
    let store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:on-focus-update',
      value: { editorId: 0 },
    });

    expect(
      editorManagerSliceKey.getSliceState(store.state)?.focusedEditorId,
    ).toBe(0);

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:on-focus-update',
      value: { editorId: 1 },
    });

    expect(
      editorManagerSliceKey.getSliceState(store.state)?.focusedEditorId,
    ).toBe(1);
  });

  test('throws error for out of bound editorId', () => {
    let store = createStore();
    expect(() => {
      store.dispatch({
        name: 'action::@bangle.io/slice-editor-manager:on-focus-update',
        value: { editorId: 1000000 },
      });
    }).toThrowError('editorId is out of range');
  });
});

describe('setting scroll position', () => {
  test('works', () => {
    let store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
      value: { wsPath: 'test:one.md', editorId: 0, scrollPosition: 9 },
    });

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toBe(9);

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
      value: { wsPath: 'test:one.md', editorId: 1, scrollPosition: 199 },
    });

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toBe(9);
    expect(geEditorScrollPosition(1, 'test:one.md')(store.state)).toBe(199);

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
      value: { wsPath: 'test:two.md', editorId: 1, scrollPosition: 299 },
    });

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toBe(9);
    expect(geEditorScrollPosition(1, 'test:one.md')(store.state)).toBe(199);
    expect(geEditorScrollPosition(1, 'test:two.md')(store.state)).toBe(299);

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
      value: { wsPath: 'test:one.md', editorId: 0, scrollPosition: 8 },
    });

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toBe(8);
    expect(geEditorScrollPosition(1, 'test:one.md')(store.state)).toBe(199);
    expect(geEditorScrollPosition(1, 'test:two.md')(store.state)).toBe(299);
  });

  test('getting non existent', () => {
    let store = createStore();

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toBe(
      undefined,
    );

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
      value: { wsPath: 'test:one.md', editorId: 0, scrollPosition: 9 },
    });

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toBe(9);
    expect(geEditorScrollPosition(1, 'test:one.md')(store.state)).toBe(
      undefined,
    );
    expect(geEditorScrollPosition(0, 'test:two.md')(store.state)).toBe(
      undefined,
    );
  });
});

describe('setting selectionJson', () => {
  const pmNode = createPMNode(
    [],
    `
# Hello
- World
    `.trim(),
  );

  test('works', () => {
    let store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(0)).toJSON(),
      },
    });

    expect(getInitialSelection(0, 'test:one.md', pmNode)(store.state))
      .toMatchInlineSnapshot(`
        Object {
          "anchor": 1,
          "head": 1,
          "type": "text",
        }
      `);

    expect(
      getInitialSelection(
        0,
        'test:one.md',
        pmNode,
      )(store.state)?.eq(Selection.near(pmNode.resolve(0))),
    ).toBe(true);

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(1)).toJSON(),
      },
    });

    expect(
      getInitialSelection(
        0,
        'test:one.md',
        pmNode,
      )(store.state)?.eq(Selection.near(pmNode.resolve(1))),
    ).toBe(true);

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:two.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(11)).toJSON(),
      },
    });

    expect(
      getInitialSelection(
        0,
        'test:two.md',
        pmNode,
      )(store.state)?.eq(Selection.near(pmNode.resolve(11))),
    ).toEqual(true);
  });

  test('getting non existent', () => {
    let store = createStore();

    expect(getInitialSelection(0, 'test:one.md', pmNode)(store.state)).toBe(
      undefined,
    );

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(11)).toJSON(),
      },
    });

    expect(
      getInitialSelection(
        0,
        'test:one.md',
        pmNode,
      )(store.state)?.eq(Selection.near(pmNode.resolve(11))),
    ).toBe(true);
    expect(getInitialSelection(1, 'test:one.md', pmNode)(store.state)).toBe(
      undefined,
    );
    expect(getInitialSelection(0, 'test:two.md', pmNode)(store.state)).toBe(
      undefined,
    );
  });
});

describe('any other action', () => {
  test('works', () => {
    let store = createStore();
    const prevState = editorManagerSliceKey.getSliceState(store.state);
    store.dispatch({
      name: 'some other action' as any,
      value: { editorId: 0 },
    });

    expect(editorManagerSliceKey.getSliceState(store.state)).toBe(prevState);
  });
});

describe('serializing state', () => {
  const pmNode = createPMNode(
    [],
    `
# Hello
- World
  `.trim(),
  );

  test('selection is serialized correctly', () => {
    let store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(5)).toJSON(),
      },
    });

    expect(
      getInitialSelection(0, 'test:one.md', pmNode)(store.state)?.toJSON(),
    ).toEqual({
      anchor: 5,
      head: 5,
      type: 'text',
    });

    const json: any = store.state.stateToJSON({
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    expect(json.editorManagerSlice.data.editorConfig.selections).toEqual([
      { 'test:one.md': Selection.near(pmNode.resolve(5)).toJSON() },
      null,
    ]);

    expect(json).toMatchSnapshot();

    const newState = AppState.stateFromJSON({
      slices: [editorManagerSlice()],
      json,
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    expect(
      getInitialSelection(0, 'test:one.md', pmNode)(newState)?.toJSON(),
    ).toEqual({
      anchor: 5,
      head: 5,
      type: 'text',
    });
  });

  test('handles schema version mismatch', () => {
    let store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(5)).toJSON(),
      },
    });

    expect(
      getInitialSelection(0, 'test:one.md', pmNode)(store.state)?.toJSON(),
    ).toEqual({
      anchor: 5,
      head: 5,
      type: 'text',
    });

    const json: any = store.state.stateToJSON({
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    json.editorManagerSlice.version = 'some-other-version';

    const newState = AppState.stateFromJSON({
      slices: [editorManagerSlice()],
      json,
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    expect(
      getInitialSelection(0, 'test:one.md', pmNode)(newState)?.toJSON(),
    ).toBe(undefined);
  });

  test('saves current editor selection correctly', () => {
    let mockEditor = createTestEditor();

    let tr = mockEditor.view.state.tr;

    mockEditor.view.dispatch(
      tr.setSelection(Selection.near(tr.doc.resolve(7))),
    );

    const store = createStore();

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor,
        editorId: 0,
      },
    });

    const json: any = store.state.stateToJSON({
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    expect(json.editorManagerSlice.data.editorConfig.selections)
      .toMatchInlineSnapshot(`
        Array [
          Object {
            "test:first.md": Object {
              "anchor": 7,
              "head": 7,
              "type": "text",
            },
          },
          null,
        ]
      `);
  });

  test('overwrites pre-existing scroll position when serializing', () => {
    let mockEditor = createTestEditor('test:one.md', `# hello world`);

    (getScrollParentElement as any).mockImplementation((): any => ({
      scrollTop: 5,
    }));

    const store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor,
        editorId: 0,
      },
    });

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-scroll-position',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        scrollPosition: 2,
      },
    });

    expect(geEditorScrollPosition(0, 'test:one.md')(store.state)).toEqual(2);

    const json: any = store.state.stateToJSON({
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    const newState = AppState.stateFromJSON({
      slices: [editorManagerSlice()],
      json,
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    // should pick up the latest position
    expect(geEditorScrollPosition(0, 'test:one.md')(newState)).toEqual(5);
  });

  test('overwrites pre-existing selections when serializing', () => {
    let mockEditor = createTestEditor('test:one.md', `# hello world`);
    const store = createStore();
    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:update-initial-selection-json',
      value: {
        wsPath: 'test:one.md',
        editorId: 0,
        selectionJson: Selection.near(pmNode.resolve(5)).toJSON(),
      },
    });

    expect(
      getInitialSelection(0, 'test:one.md', pmNode)(store.state)?.toJSON(),
    ).toEqual({
      anchor: 5,
      head: 5,
      type: 'text',
    });

    let tr = mockEditor.view.state.tr;

    mockEditor.view.dispatch(
      tr.setSelection(Selection.near(tr.doc.resolve(2))),
    );

    store.dispatch({
      name: 'action::@bangle.io/slice-editor-manager:set-editor',
      value: {
        editor: mockEditor,
        editorId: 0,
      },
    });

    const json: any = store.state.stateToJSON({
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    const newState = AppState.stateFromJSON({
      slices: [editorManagerSlice()],
      json,
      sliceFields: { editorManagerSlice: editorManagerSlice() },
    });

    expect(
      getInitialSelection(0, 'test:one.md', pmNode)(newState)?.toJSON(),
    ).toEqual(Selection.near(pmNode.resolve(2)).toJSON());
  });
});
