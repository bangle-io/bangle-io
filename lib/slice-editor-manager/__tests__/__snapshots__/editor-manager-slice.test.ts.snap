// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializing state selection is serialized correctly 1`] = `
Object {
  "editorManagerSlice": Object {
    "data": Object {
      "editorConfig": Object {
        "scrollPositions": Array [
          null,
          null,
          null,
        ],
        "selections": Array [
          Object {
            "test:one.md": Object {
              "anchor": 5,
              "head": 5,
              "type": "text",
            },
          },
          null,
          null,
        ],
      },
      "focusedEditorId": undefined,
    },
    "version": "editor-slice/2",
  },
}
`;
